File: MapXtreme/6.6/Samples/VisualStudio2005/Web/Features/HelloWorld/HelloWorldCS/App_Code/NorthArrowAdornment.cs
using System;
using System.Drawing;
using System.Runtime.Serialization;
using MapInfo.Mapping;

namespace HelloWorldWeb
{
	[Serializable]
	public sealed class NorthArrowAdornment : MapInfo.Mapping.IAdornment, ISerializable
	{
		private System.Drawing.Size _size;
		private System.Drawing.Point _location;
		private System.String _name;
		private System.String _alias;
		private System.String _mapAlias;
		private System.IO.MemoryStream _imageStream;
		private event ViewChangedEventHandler _viewChangedEvent=null;

		private bool _blockViewChangedEvent = false;

		public NorthArrowAdornment(string mapAlias, Size size, string adornmentAlias, string name, string fileName)
		{
			_size = size;
			_alias = adornmentAlias;
			_name = name;
			_location = new System.Drawing.Point(0, 0);
			_mapAlias = mapAlias;

			_imageStream = new System.IO.MemoryStream();
			System.Drawing.Image image = System.Drawing.Imaging.Metafile.FromFile(fileName);
			image.Save(_imageStream, image.RawFormat);
			_imageStream.Position = 0;
			image.Dispose();
		}

		/// <summary>
		/// Internal use only.
		/// </summary>
		/// <param name="adornmentAlias">alias name for this adornment.</param>
		/// <param name="mapAlias">Map alias name.</param>
		public NorthArrowAdornment(string adornmentAlias, string mapAlias)
		{
			_alias = adornmentAlias;
			_mapAlias = mapAlias;
		}

		#region IAdornment Members
		public System.Drawing.Size Size
		{
			get
			{ 
				return _size;
			}
			set
			{
				if (_size != value) 
				{
					_size = value;
					OnViewSizeChanged();
				}
			}
		}

		public string Name
		{
			get
			{
				return _name;
			}
			set
			{
				_name = value;
			}
		}

		public System.Drawing.Point Location
		{
			get
			{
				return _location;
			}
			set
			{
				if(this.BlockViewChangedEvent == false)
				{
					if(_location != value)
					{
						_location = value;
						OnViewLocationChanged();
					}
				}
			}
		}

		public string Alias
		{
			get
			{
				return _alias;
			}
			set
			{
				if (this.Map != null) 
				{
					// check if setting alias back to same value
					// or if alias is not unique
					IAdornment a = this.Map.Adornments[value];
					if (a != null && !object.ReferenceEquals(this, a))
					{
						throw new ArgumentException("AliasNotUnique!!!");
					}
				}
				_alias = value;
			}
		}

		public MapInfo.Mapping.Map Map
		{
			get
			{
				return MapInfo.Engine.Session.Current.MapFactory[_mapAlias];
			}
		}

		public void Draw(System.Drawing.Graphics graphics, System.Drawing.Rectangle updateArea, System.Drawing.Point drawPnt)
		{
			_imageStream.Position = 0;
			System.Drawing.Image image = System.Drawing.Imaging.Metafile.FromStream(_imageStream);
			graphics.DrawImage(image, drawPnt);
			image.Dispose();
		}

		public event ViewChangedEventHandler ViewChangedEvent
		{
			add
			{
				_viewChangedEvent += value;
			}
			remove
			{
				_viewChangedEvent -= value;
			}
		}
		#endregion

		public bool BlockViewChangedEvent
		{
			get
			{
				return _blockViewChangedEvent;
			}
			set
			{
				_blockViewChangedEvent = value;
			}
		}

		private void OnViewChanged(ViewChangedEventArgs e) 
		{
			if(!BlockViewChangedEvent)
			{
				_viewChangedEvent(this, e);
			}
		}

		private void OnViewChanged() 
		{
			ViewChangedEventArgs e=new ViewChangedEventArgs();
			OnViewChanged(e);
		}

		private void OnViewSizeChanged() 
		{
			ViewChangedEventArgs e=new ViewChangedEventArgs();
			e.SizeChange = true;
			OnViewChanged(e);
		}

		private void OnViewLocationChanged() 
		{
			ViewChangedEventArgs e=new ViewChangedEventArgs();
			e.LocationChange = true;
			OnViewChanged(e);
		}

		void ISerializable.GetObjectData(SerializationInfo info, StreamingContext context)
		{
			info.AddValue("_alias", _alias);
			info.AddValue("_name", _name);
			info.AddValue("_size", _size);
			info.AddValue("_mapAlias", _mapAlias);
			info.AddValue("_location", _location);
			info.AddValue("_imageStream", _imageStream);
			info.SetType(typeof(NorthArrowAdornmentDeserializer));
		}

		public void SetObjectData(SerializationInfo info, StreamingContext context)
		{
			_name = info.GetString("_name");
			_size = (System.Drawing.Size)info.GetValue("_size", typeof(System.Drawing.Size));
			_alias = info.GetString("_alias");
			_mapAlias = info.GetString("_mapAlias");
			_location = (System.Drawing.Point)info.GetValue("_location", typeof(System.Drawing.Point));
			_imageStream = (System.IO.MemoryStream)info.GetValue("_imageStream", typeof(System.IO.MemoryStream));
		}
	}
	/// <summary>Implements deserialization of a NorthArrowAdornment instance.</summary>
	/// <remarks>See the article at 
	/// http://msdn.microsoft.com/msdnmag/issues/02/07/net/print.asp.
	/// See also the "Object References" section of the article at
	/// http://community.borland.com/article/0,1410,29787,00.html.
	/// </remarks>
	[Serializable]
	public sealed class NorthArrowAdornmentDeserializer : IObjectReference, ISerializable
	{
		NorthArrowAdornment _adornment = null;

		private NorthArrowAdornmentDeserializer(SerializationInfo info, StreamingContext context)
		{
			string alias = info.GetString("_alias");
			string mapAlias = info.GetString("_mapAlias");

			// See if the "same" map exists.
			MapInfo.Mapping.Map map = MapInfo.Engine.Session.Current.MapFactory[mapAlias];
			if (map == null)
			{
				throw new SerializationException(MapInfo.Engine.Session.Current.Resources.GetString("MapInfo.Serialization.AdornmentMapNotFound", alias, mapAlias));
			}
			
			IAdornment adornment = map.Adornments[alias];
			if (adornment != null)
			{
				if (adornment is NorthArrowAdornment) 
				{
					_adornment = (NorthArrowAdornment)adornment;
				}
				else
				{
					map.Adornments.Remove(alias);
					_adornment=null;
					adornment=null;
				}
			}
			// The "same" adornment doesn't exist. Create it from scratch.
			if (adornment == null ) 
			{
				_adornment = new NorthArrowAdornment(alias, mapAlias);
				//adornment = _adornment;
				map.Adornments.Append(_adornment);
			}

			_adornment.SetObjectData(info, context);
		}
		Object IObjectReference.GetRealObject(StreamingContext context)
		{
			// NOTE: This method is called twice for each deserialization.
			// According to the http://community.borland.com/article/0,1410,29787,00.html
			// article, this was known by Microsoft as a bug in the 1.1 beta.
			// There is a work around, but as long as
			// there's no processing done here, no harm is done.
			return _adornment;
		}
		///	<summary>Required by the ISerializable interface, but this one is never called.</summary>
		///	<param name="info">The SerializationInfo to populate with data.</param>
		///	<param name="context">The destination for this serialization.</param>
		void ISerializable.GetObjectData(SerializationInfo info, StreamingContext context) { }
	}
}

File: trunk/MapXtreme/6.6/Samples/VisualStudio2005/Web/Features/HelloWorld/HelloWorldVB/App_Code/NorthArrowAdornment.vb
Imports System.IO
Imports System.Drawing
Imports System.Runtime.Serialization


Namespace HelloWorldWeb

<Serializable()> _
Public NotInheritable Class NorthArrowAdornment
    Implements MapInfo.Mapping.IAdornment, ISerializable
    ' Fields
    Private _alias As String
    Private _imageStream As System.IO.MemoryStream
    Private _location As System.Drawing.Point
    Private _mapAlias As String
    Private _name As String
    Private _size As System.Drawing.Size
    Private _blockViewChangedEvent As Boolean = False

    Public Sub New(ByVal mapAlias As String, ByVal size As System.Drawing.Size, ByVal adornmentAlias As String, ByVal name As String, ByVal fileName As String)
        Me._size = size
        Me._alias = adornmentAlias
        Me._name = name
        Me._location = New Point(0, 0)
        Me._mapAlias = mapAlias
        Me._imageStream = New System.IO.MemoryStream
        Dim image As Image = image.FromFile(fileName)
        image.Save(Me._imageStream, image.RawFormat)
        Me._imageStream.Position = 0
        image.Dispose()
    End Sub

    Public Sub New(ByVal adornmentAlias As String, ByVal mapAlias As String)
        Me._alias = adornmentAlias
        Me._mapAlias = mapAlias
    End Sub
#Region "IAdornment Members"
    Public Property Size() As System.Drawing.Size Implements MapInfo.Mapping.IAdornment.Size
        Get
            Return _size
        End Get
        Set(ByVal value As Size)
            If (Not _size.Equals(value)) Then
                Me._size = value
                Me.OnViewSizeChanged()
            End If
        End Set
    End Property
    Public Property Name() As String Implements MapInfo.Mapping.IAdornment.Name
        Get
            Return _name
        End Get
        Set(ByVal value As String)
            _name = value
        End Set
    End Property

    Public Property Location() As System.Drawing.Point Implements MapInfo.Mapping.IAdornment.Location
        Get
            Return _location
        End Get
        Set(ByVal value As System.Drawing.Point)
            If (Not Me.BlockViewChangedEvent AndAlso (Not _location.Equals(value))) Then
                _location = value
                OnViewLocationChanged()
            End If
        End Set
    End Property

    Public Property [Alias]() As String Implements MapInfo.Mapping.IAdornment.Alias
        Get
            Return _alias
        End Get
        Set(ByVal value As String)
            If (Not Me.Map Is Nothing) Then
                Dim adornment As MapInfo.Mapping.IAdornment = Me.Map.Adornments.Item(value)
                If ((Not adornment Is Nothing) AndAlso Not Object.ReferenceEquals(Me, adornment)) Then
                    Throw New ArgumentException("AliasNotUnique!!!")
                End If
            End If
            _alias = value
        End Set
    End Property

    Public ReadOnly Property Map() As MapInfo.Mapping.Map Implements MapInfo.Mapping.IAdornment.Map
        Get
            Return MapInfo.Engine.Session.Current.MapFactory.Item(_mapAlias)
        End Get
    End Property


    Public Sub Draw(ByVal graphics As System.Drawing.Graphics, ByVal updateArea As System.Drawing.Rectangle, ByVal drawPnt As System.Drawing.Point) Implements MapInfo.Mapping.IAdornment.Draw
        _imageStream.Position = 0
        Dim image As System.Drawing.Image = System.Drawing.Image.FromStream(Me._imageStream)
        graphics.DrawImage(image, drawPnt)
        image.Dispose()
    End Sub

    Public Event ViewChangedEvent(ByVal sender As Object, ByVal e As MapInfo.Mapping.ViewChangedEventArgs) Implements MapInfo.Mapping.IAdornment.ViewChangedEvent

#End Region

    Public Property BlockViewChangedEvent() As Boolean
        Get
            Return _blockViewChangedEvent
        End Get
        Set(ByVal value As Boolean)
            _blockViewChangedEvent = value
        End Set
    End Property

    Private Sub OnViewChanged(ByVal e As MapInfo.Mapping.ViewChangedEventArgs)
        If Not Me.BlockViewChangedEvent Then
            RaiseEvent ViewChangedEvent(Me, e)
        End If
    End Sub

    Private Sub OnViewChanged()
        Dim e As New MapInfo.Mapping.ViewChangedEventArgs
        Me.OnViewChanged(e)
    End Sub

    Private Sub OnViewSizeChanged()
        Dim e As New MapInfo.Mapping.ViewChangedEventArgs
        e.SizeChange = True
        Me.OnViewChanged(e)
    End Sub

    Private Sub OnViewLocationChanged()
        Dim e As New MapInfo.Mapping.ViewChangedEventArgs
        e.LocationChange = True
        Me.OnViewChanged(e)
    End Sub

    Private Sub GetObjectData(ByVal info As SerializationInfo, ByVal context As StreamingContext) Implements ISerializable.GetObjectData
        info.AddValue("_alias", Me._alias)
        info.AddValue("_name", Me._name)
        info.AddValue("_size", Me._size)
        info.AddValue("_mapAlias", Me._mapAlias)
        info.AddValue("_location", Me._location)
        info.AddValue("_imageStream", Me._imageStream)
        info.SetType(GetType(NorthArrowAdornmentDeserializer))
    End Sub

    Public Sub SetObjectData(ByVal info As SerializationInfo, ByVal context As StreamingContext)
        Me._name = info.GetString("_name")
        Me._size = CType(info.GetValue("_size", GetType(Size)), Size)
        Me._alias = info.GetString("_alias")
        Me._mapAlias = info.GetString("_mapAlias")
        Me._location = CType(info.GetValue("_location", GetType(Point)), Point)
        Me._imageStream = CType(info.GetValue("_imageStream", GetType(MemoryStream)), MemoryStream)
    End Sub
End Class

' <summary>Implements deserialization of a NorthArrowAdornment instance.</summary>
' <remarks>See the article at 
' http://msdn.microsoft.com/msdnmag/issues/02/07/net/print.asp.
' See also the "Object References" section of the article at
' http://community.borland.com/article/0,1410,29787,00.html.
' </remarks>
<Serializable()> _
Public NotInheritable Class NorthArrowAdornmentDeserializer
    Implements IObjectReference, ISerializable

    ' Fields
    Private _adornment As NorthArrowAdornment = Nothing

    Private Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)
        Dim [alias] As String = info.GetString("_alias")
        Dim mapAlias As String = info.GetString("_mapAlias")

        ' See if the "same" map exists.
        Dim map As MapInfo.Mapping.Map = MapInfo.Engine.Session.Current.MapFactory.Item(mapAlias)
        If (map Is Nothing) Then
            Throw New SerializationException(MapInfo.Engine.Session.Current.Resources.GetString("MapInfo.Serialization.AdornmentMapNotFound", [alias], mapAlias))
        End If

        Dim adornment As MapInfo.Mapping.IAdornment = map.Adornments.Item([alias])
        If (Not adornment Is Nothing) Then
            If TypeOf adornment Is NorthArrowAdornment Then
                _adornment = CType(adornment, NorthArrowAdornment)
            Else
                map.Adornments.Remove([alias])
                _adornment = Nothing
                adornment = Nothing
            End If
        End If
        ' The "same" adornment doesn't exist. Create it from scratch.
        If (adornment Is Nothing) Then
            _adornment = New NorthArrowAdornment([alias], mapAlias)
            map.Adornments.Append(_adornment)
        End If
        _adornment.SetObjectData(info, context)
    End Sub


    Private Function GetRealObject(ByVal context As StreamingContext) As Object Implements IObjectReference.GetRealObject
        ' NOTE: This method is called twice for each deserialization.
        ' According to the http://community.borland.com/article/0,1410,29787,00.html
        ' article, this was known by Microsoft as a bug in the 1.1 beta.
        ' There is a work around, but as long as
        ' there's no processing done here, no harm is done.
        Return _adornment
    End Function

    ' <summary>Required by the ISerializable interface, but this one is never called.</summary>
    ' <param name="info">The SerializationInfo to populate with data.</param>
    ' <param name="context">The destination for this serialization.</param>
    Private Sub GetObjectData(ByVal info As SerializationInfo, ByVal context As StreamingContext) Implements ISerializable.GetObjectData
    End Sub
End Class

End Namespace

